### ISSUE TITLE
Traefik doesn't auto re-add re-enabled service based on Consul Catalog changes

### ISSUE THREAD

**Issue by xmackex** [containous/traefik#3157]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

- register service -> service comes available in traefik
- stop service -> service is gone from traefik
- start same service -> service is not available in traefik but alive in Consul
- wait 5-20 minutes -> service is not available in traefik
- restart traefik -> service comes available in traefik

### What did you expect to see?

It seems to be a sub-issue of [previously closed issue](https://github.com/containous/traefik/issues/2552)

### What did you see instead?

Changes are properly updated, but in case of deregistering and registering same service there are issues.

Also I believe I could misconfigured something, would appreciate for hints

### Output of `traefik version`: (_What version of Traefik are you using?_)

Consul version: 1.0.6
Traefik version: v1.5.4, v1.6.0-rc4

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

My Traefik config is:

```toml
[entryPoints]
  [entryPoints.http]
    address = ":5555"
    compress = false
    [entryPoints.http.proxyProtocol]
        trustedIPs = ["10.0.0.0/8"]
  [entryPoints.traefik_api]
    address = ":1111"
    compress = false
    
[ping]
  entryPoint = "http"

[consulCatalog]
endpoint = "127.0.0.1:8450"
exposedByDefault = false
prefix = "traefik"

[api]
  entryPoint = "traefik_api"
  dashboard = true
  debug = true
[traefikLog]
  format   = "json"
[accessLog]
  format   = "json"
defaultEntryPoints = ["http"]
```

Frontend rule is exposed via service tag: `traefik.frontend.rule=Host:example.com`

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of communication regulation.

Thanks for your interest in Træfik!

Please edit your issue and use the [provided template](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE.md).

Issue templates help us help you by providing all necessary information.

Here are the available templates:

- [Standard issue](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE.md)
- [Feature request](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE/features.md)
- [Bug report](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE/bugs.md)

And remember: each time someone ignores the template, a cute little bunny dies.
<details>
<summary>#SaveTheCuteBunny :heart: :rabbit: :heart:</summary>

![southpark - estrella kill rabbit](https://user-images.githubusercontent.com/5674651/27143914-3cea7896-5130-11e7-81f8-02d8ac2da7a7.gif)

![southpark - estrella kill rabbit2](https://user-images.githubusercontent.com/5674651/27143924-49104c72-5130-11e7-82ae-ccedbd5dfa1c.gif)

</details>

**Comment by mmatur** [comment]
Hi @xmackex,

Thanks for your interest in Træfik.

If it's possible, could you please send us a reproducible stack. It will help us a lot to reproduce your issue. 

We need to understand how your services are registered and deregistered from consul, if you have health-check configured on your services etc.

**Comment by xmackex** [comment]
Hi @mmatur 

Service in consul is registered via nomad,
I have 2 configured health-checks:
1. HTTP check to `health` endpoint
2. Simple process check via `cat /proc/1/stat`

Reproduce stack:
1. `nomad run service.nomad`
2. `nomad stop -purge service.nomad`
3. `nomad run service.nomad`

Service Nomad is not changed

Please let me know if more info required.

**Comment by imcitius** [comment]
Hello

I can confirm the same issue.

My services is deployed to virtual machines, and is registered in consul, using simple json config with service description, traefik tags, and some health checks, for example:

```
{
  "services": [
    {
      "name": "backend",
      "address": "10.20.30.1",
      "port": 80,
      "tags": [
        "traefik.frontend.rule=Host:backend.app.com",
        "traefik.frontend.passHostHeader=true",
        "traefik.frontend.entryPoints=http,https",
        "traefik.backend.loadbalancer.stickiness=true",
        "traefik.protocol=http",
        "traefik.enable=true",
        "backend"
      ],
      "checks": [
        {
          "name": "nginx is running",
          "id": "nginx-run",
          "args": ["systemctl", "status", "openresty"],
          "interval": "10s",
          "timeout": "2s"
        },
        {
          "name": "nginx is listening http",
          "id": "nginx-http-listen",
          "tcp": "{{ ansible_default_ipv4.address }}:80",
          "interval": "10s",
          "timeout": "2s"
        },
        {
          "name": "php-fpm is running",
          "id": "php-fpm-run",
          "args": ["systemctl", "status", "php-fpm"],
          "interval": "10s",
          "timeout": "2s"
        },
        {
          "id": "api",
          "name": "HTTP API healthcheck",
          "http": "http://localhost/monitor/status.php",
          "tls_skip_verify": true,
          "method": "GET",
          "interval": "10s",
          "timeout": "3s"
        }
      ]
    }
```

**Comment by trondhindenes** [comment]
Just out of curiosity: Is the Consul environment locked down with strict ACLs? I've seen an issue where Traefik didn't have enough permissions to put a "watch" on the service in question. Baed on that experience it would probably be valuable info to @mmatur to know if this is reproducable in an environment without locked-down ACLs.

**Comment by mmatur** [comment]
Hi @xmackex, 

Thanks for these informations. After some tests with Consul, Nomad and Træfik I am not able to reproduce your issue. 

Could you give me more informations about your Nomad service, your Consul configuration and how you start Træfik.

**Comment by xmackex** [comment]
Hi @mmatur , one thing I forgot to mention,
Consul is running under strict ACL
Token policy:
```
node "" { policy = "read" } service "" { policy = "read" } agent "" { policy = "read" }
```
Basically this policy should allow read all.
Switched token to "management" type , but same result.

Here is some log with level=DEBUG
After service purge:
```
{"MissingServices":["_nomad-task-4z7hdz2nftwncfj24ohjns6gndx37ro7"],"level":"debug","msg":"Health State change detected.","time":"2018-04-14T14:27:14Z"}
{"level":"debug","msg":"List of services changed","time":"2018-04-14T14:27:14Z"}
```

After service re-enable:
```
{"level":"debug","msg":"List of services changed","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"consul","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service consul","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"fabio-pub","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service fabio-pub","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"nomad-client","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service nomad-client","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"nomad-server","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service nomad-server","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service nomad-server","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service nomad-server","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"test-web-service","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Configuration received from provider consul_catalog: {}","time":"2018-04-14T14:28:49Z"}
{"level":"info","msg":"Skipping same configuration for provider consul_catalog","time":"2018-04-14T14:28:49Z"}
{"DiscoveredServices":["_nomad-task-2btqqx7kuhgpicprjhwdkg4hxr6zbohr"],"level":"debug","msg":"Health State change detected.","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"List of services changed","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"fabio-pub","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service fabio-pub","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"nomad-client","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service nomad-client","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"nomad-server","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service nomad-server","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service nomad-server","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service nomad-server","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"test-web-service","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Fetching service","service":"consul","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Filtering disabled Consul service consul","time":"2018-04-14T14:28:49Z"}
{"level":"debug","msg":"Configuration received from provider consul_catalog: {}","time":"2018-04-14T14:28:49Z"}
{"level":"info","msg":"Skipping same configuration for provider consul_catalog","time":"2018-04-14T14:28:49Z"}
```

But unfortunately I can't see new service in dashboard.

**Comment by trondhindenes** [comment]
This is definetely an issue, specifically testing on 1.6.0 now. For single-node services, they don't come back consistently after being maint-moded. We're going live with our heaviest api behind Traefik this week, and are forced to write a "helper service" which bumps things back into shape when Traefik loses sync. Pretty bad :-(

**Comment by thebinary** [comment]
i find two related issues: 
1. If you re-enable agent that is registering the backend service, the backend service (that is registered by that particular agent) is not re-enabled in traefik.
2. If the health-checks state change for a backend service, the change is not reflected in traefik.

which issue are we discussing in this thread?

**Comment by thebinary** [comment]
Traefik watches for change in "Service" endpoint of consul catalog and triggers a function to find healthy nodes immediately. So, the newly registered service node may not be healthy yet when traefik does this query. So, it misses to add the new node even after detecting a change in service.

I have tested a solution by adding a sleep timer before traefik does this query, to let the service come online, and it WORKS!! But this is not a PROPER SOLUTION.

**Comment by mmatur** [comment]
@thebinary @trondhindenes @xmackex after a lot of test on this issue. I am really not able to reproduce your issue. Could you please give me some hints to be able to reproduce this issue easily?

**Comment by thebinary** [comment]
@mmatur configure a service with health checks that take a bit long time (for being sure to reproduce the issue). Also, make sure you DONT HAVE "status": "passing" set for the health check. I think with this you can reprodue the issue.

**Comment by thebinary** [comment]
@ldez i dont think issue #3380 is a duplicate of this issue. There are two related issues but quite different:
1. Issue with re-registration of a service.
2. Issue with reflection of healthy/unhealthy nodes (health checks in consul catalog)

I think this issue thread for the 1st. And I opened #3380 for the 2nd.

**Comment by mmatur** [comment]
Hi @thebinary,

Thanks for your feedback. 

**With Passing status**

```bash
curl -X PUT \
  http://127.0.0.1:8500/v1/agent/service/register \
  -H 'Content-Type: application/json' \
  -d '{
    "name": "my-api",
    "tags": [
        "api",
        "traefik.enable=true"
    ],
    "port": 4242,
    "checks": [
       {
         "http": "http://127.0.0.1:4242/health",
         "interval": "5s",
         "timeout": "5s",
         "service_id": "my-api",
         "status": "passing"
       }
    ]
}'
```

**Without passing status**

```bash
curl -X PUT \
  http://127.0.0.1:8500/v1/agent/service/register \
  -H 'Content-Type: application/json' \
  -d '{
    "name": "my-api",
    "tags": [
        "api",
        "traefik.enable=true"
    ],
    "port": 4242,
    "checks": [
       {
         "http": "http://127.0.0.1:4242/health",
         "interval": "5s",
         "timeout": "5s",
         "service_id": "my-api",
       }
    ]
}'
```

I have exactly the same result in Træfik. I am not able to reproduce the issue.

**Comment by spunti** [comment]
Hi,

i´m faced with the same bug as reported here #3380 
I have to agree with @thebinary that this are different issues.

@mmatur did you created more than one host with your service? So that in traefik more than one server is listed at the backend section? Because if there is only one server and this one gets unhealthy, traefik is working like expected and removes the complete backend immediately.

**Comment by thebinary** [comment]
@spunti yes i agree with you. Traefik works like expected for only one server. 

Actually it works as expected in following conditions 
1. All servers for a service get unhealthy
2. One or more server get healthy when previously all servers were unhealthy
I have mentioned this at the issue #3380 

And when only one server is involved, in all cases the above two conditions match thats why it works.

**Comment by thebinary** [comment]
@mmatur try adding a dummy sleep of 5-10s in your '/health' check endpoint before returning 200 OK. and please register the service from at least two nodes/servers, and then try de-registering and re-registering one of the nodes. I hope you will have reproduced the issue then.

**Comment by mmatur** [comment]
@spunti @thebinary Thanks for tips. I am able to reproduce the issue with health check.

@xmackex Could you please send us more information about your nomad service you are using please ?

**Comment by thebinary** [comment]
I just came to know PR #3390 fixes this issue as well. Though the new new node is not added from the Service Watch endpoint as discussed in this thread. A bit later it will be added because of the health state change detection which has been solved by #3390

**Comment by xmackex** [comment]
@mmatur sorry for really late reply, unfortunately had bunch of stuff to do.
So via Nomad I'm running simple apache with PHP and 2 healthchecks (1 HTTP, 1 command)

Service block looks like:

```hcl
      service {
        name = "SOME_NAME-SOME_ENV"
        tags = [
          "urlprefix_pub-stg.PROJECT.example.com/",
          "traefik.enable=true",
          "traefik.frontend.rule=Host:stg.PROJECT.example.com",
        ]
        address_mode = "host"
        port = "http"
        check {
          name     = "SOME_NAME-SOME_ENV-http"
          type     = "http"   # possible script, http, tcp
          protocol = "http"
          path    = "/healthcheck.php"
          port    = "http"
          interval = "5s"
          timeout  = "3s"    # timeout_http
        }
        check {
          name     = "SOME_NAME-SOME_ENV-process"
          type     = "script"   # possible script, http, tcp
          command = "cat"
          args = [
            "/proc/1/stat"
          ]
          interval = "5s"
          timeout  = "1s"    # timeout_process
        }
        check_restart {
          limit = 2
          grace  = "5s"   # timeout_http
          ignore_warnings = false
        }
      }
```

**Comment by trondhindenes** [comment]
We just took a prolonged downtime on our main website because of this error. We remove nodes from Consul (using maint) pre deploy, and add them after deploy. It seems to me that Traefik doesn't reliably pick up re-added servers. We have a helper-script which bumps consul when there's a diff, but for Traefik 1.6.2 it doesnt seem to do a "full refresh" of the catalog anymore whenever we restart the consul service - which means that we're essentially unable to get things working without restarting the Traefik service itself (which we don't want to do). Today was not a good day do be a Traefik user.

**Comment by trondhindenes** [comment]
awesome, will deploy to our staging env on monday and test it out!

**Comment by trondhindenes** [comment]
@mmatur the fix looks good - verified in prod aswell. Thanks!

**Comment by trondhindenes** [comment]
That said, it still doesn't seem to react to "maintenance mode" only - the service has to actually fail for Traefik to pick it up. Is that by design?

**Comment by trondhindenes** [comment]
I was too quick. The service does not re-add servers that were previously maintenance-moded. I'm afraid we're forced to move off Traefik if doesn't get fixed, as we can't reliably deploy services right now.

**Comment by trondhindenes** [comment]
Consul:
![screenshot from 2018-06-04 13-42-33](https://user-images.githubusercontent.com/1747120/40915676-3e31c1ca-67fd-11e8-8d0a-53c770437c14.png)

Traefik:
![screenshot from 2018-06-04 13-43-34](https://user-images.githubusercontent.com/1747120/40915687-4cdbd490-67fd-11e8-92bd-ede49ebee5de.png)

The "redirect" service was put in "maintenance-mode" (each node was brought out and then in) and now it just sits there missing a server. I get the feeling that if Traefik misses an "event" from Consul it never gets into sync again.

**Comment by traefiker** [comment]
Closed by #3434.
